rust/src/two_stacks/mod.rs

```diff
@@ -0,0 +1,73 @@
+use crate::FifoWindow;
+use alga::general::AbstractMonoid;
+use alga::general::Operator;
+use std::marker::PhantomData;
+
+#[derive(Debug)]
+struct Item<Value>
+where
+    Value: Clone,
+{
+    agg: Value,
+    val: Value,
+}
+
+#[derive(Debug)]
+pub struct TwoStacks<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    front: Vec<Item<Value>>,
+    back: Vec<Item<Value>>,
+    op: PhantomData<BinOp>,
+}
+
+impl<Value, BinOp> FifoWindow<Value, BinOp> for TwoStacks<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    fn new() -> TwoStacks<Value, BinOp> {
+        TwoStacks {
+            front: Vec::new(),
+            back: Vec::new(),
+            op: PhantomData,
+        }
+    }
+    fn push(&mut self, v: Value) {
+        self.back.push(Item {
+            agg: Self::agg(&self.back).operate(&v),
+            val: v,
+        });
+    }
+    fn pop(&mut self) {
+        if self.front.is_empty() {
+            while let Some(top) = self.back.pop() {
+                self.front.push(Item {
+                    agg: top.val.operate(&Self::agg(&self.front)),
+                    val: top.val,
+                })
+            }
+        }
+        self.front.pop();
+    }
+    fn query(&self) -> Value {
+        Self::agg(&self.front).operate(&Self::agg(&self.back))
```

Comment:
Instead of defining and calling the `agg` function, is it possible to instead call something like `stack.last().unwrap_or(Value::identity())`? Or no because we also need to call `clone()`?

Reply:
It is possible. I can change it if you want, I think it would look like this:
```rust
     fn query(&self) -> Value {
         self.front
             .last()
             .map(|item| item.agg.clone())
             .unwrap_or_else(|| Value::identity())
             .operate(
                 &self
                     .back
                     .last()
                     .map(|item| item.agg.clone())
                     .unwrap_or_else(|| Value::identity()),
             )
     }
```
I think I defined it as a separate function to make it look like your definition in *Low-Latency Sliding-Window Aggregation in Worst-Case Constant Time*.

Reply:
You're right, I think your original implementation is better. Having to duplicate the `clone()` and the optional handling with `Value::identity()` takes away from the simplicity of the operation. Thanks for indulging my question!